app/gui2/src/util/ast/abstract.ts

```diff
@@ -0,0 +1,924 @@
+function getNode(id: AstId): Ast | null {
+  return committed.get(id) ?? null
+}
+
+/** Returns a syntax node representing the current state of the given ID, including any uncommitted modifications. */
+function getUncommitted(id: AstId): Ast | null {
+  if (edited.has(id)) {
+    return edited.get(id) ?? null
+  } else {
+    return committed.get(id) ?? null
+  }
+}
+
+declare const brandAstId: unique symbol
+declare const brandTokenId: unique symbol
+export type AstId = ExprId & { [brandAstId]: never }
+export type TokenId = ExprId & { [brandTokenId]: never }
+
+function newNodeId(): AstId {
+  return random.uuidv4() as AstId
+}
+function newTokenId(): TokenId {
+  return random.uuidv4() as TokenId
+}
+
+export type NodeChild = { whitespace?: string | undefined; node: AstId | Tok }
+type AstWithWhitespace = { whitespace?: string | undefined; node: AstId }
+type TokWithWhitespace = { whitespace?: string | undefined; node: Tok }
+
+function token(code: string, type: Token.Type): Tok {
+  return new Tok(code, newTokenId(), type)
+}
+
+export function isToken(node: Tok | AstId): node is Tok {
+  return node instanceof Tok
+}
+
+export abstract class Expression {
+  abstract code(): string
+
+  /** Returns child subtrees, without information about the whitespace between them. */
+  *children(): IterableIterator<Ast | Tok> {
+    for (const child of this._rawChildren()) {
+      if (child.node instanceof Tok) {
+        yield child.node
+      } else {
+        const node = getNode(child.node)
+        if (node) yield node
+      }
+    }
+  }
+
+  /** Returns child concrete subtrees. */
+  abstract _rawChildren(): IterableIterator<NodeChild>
+}
+
+export class Tok extends Expression {
+  private _code: string
+  exprId: TokenId
+  readonly _tokenType: Token.Type
+  constructor(code: string, id: TokenId, type: Token.Type) {
+    super()
+    this._code = code
+    this.exprId = id
+    this._tokenType = type
+  }
+
+  code(): string {
+    return this._code
+  }
+
+  *_rawChildren(): IterableIterator<NodeChild> {}
+
+  typeName(): string {
+    return Token.typeNames[this._tokenType]!
+  }
+}
+
+type Span = { start: number; end: number; whitespaceLength: number }
+export abstract class Ast extends Expression {
+  readonly treeType: Tree.Type | undefined
+  _id: AstId
+  // TODO: Eliminate this before enabling edit support.
+  readonly span: Span
+
+  get exprId(): AstId {
+    return this._id
+  }
+
+  code(): string {
+    return this._print().code
+  }
+
+  typeName(): string | undefined {
+    if (this.treeType === undefined) return undefined
+    return Tree.typeNames[this.treeType]
+  }
+
+  static parse(source: PrintedSource | string): Ast {
+    const code = typeof source === 'object' ? source.code : source
+    const ids = typeof source === 'object' ? source.info : undefined
+    const tree = parseEnso(code)
+    return getUncommitted(abstract(tree, code, ids).node)!
+  }
```

Comment:
Initially, I had no idea why this code works. As I see, `abstract` does not set any of `edited` or `commited`. Realizing where actually globals are affected requires some investigation.

In general, this "global" approach a bit eccentric. Why do we need this? Is it some sort of optimization? What if I would like to parse two different codes which may share ID? What if someone modifies `abstract` code so it won't call any constructor? Not to mention, we resign this way from JS GC and must clear the memory by hand.

Even if we keep this, the code should be heavy of comments: here we need an explanation why `!` is safe.; `abstract` has to have all guarantees documented and ensured by tests.